floris/tools/optimization/scipy/derive_downstream_turbines.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2021 NREL
+
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not
+# use this file except in compliance with the License. You may obtain a copy of
+# the License at http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+
+# See https://floris.readthedocs.io for documentation
+
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+def derive_downstream_turbines(fi, wind_direction, wake_slope=0.10, plot_lines=False):
```

Comment:
I think this could work to approximate identifying downstream turbines, and as you say is meant to be conservative. But there is a similar feature in `FlowField.calculate_wake` where a list containing the number of wakes impacting each turbine can be saved. The number of wakes is determined by whether the area of overlap between each upstream wake and the turbine rotor area is greater than zero. `FlowField.calculate_wake` could be easily modified to also save a list of downstream turbines (or a list containing the number of other turbines each turbine wakes). One advantage of this approach would be that the wake boundary used to determine waked turbines would automatically adjust to different TI values and other parameters. Users wouldn't have to know what wake_slope value to use. Curious to hear your thoughts about this.

Reply:
I had thought about this too. Though, I was hesitant to do so because:
- Evaluating FLORIS is slower, though probably negligible compared to optimization time
- Sensitive situations can arise with FLORIS: the downstream turbines without wake steering can differ from which turbines are downstream with wake steering, e.g., due to secondary steering effects or changes in turbulence intensity
- To be conservative, one can evaluate FLORIS with a high TI/large wd spread and derive it from that, but I feel like the current simplified model will always be conservative with the default wake_slope value. The user really need not change it unless the user is aiming to be less conservative.

I like your idea too, and it would circumvent this problem of having an entirely different wake model to determine which turbines are downstream. Can you point me to the line in calculate_wake where you envision the additional functionality to be implemented?

Reply:
Here is the line where the number of upstream turbines waking each turbine gets tracked in flow_field.calculate_wake. This is where another variable could track the number of downstream turbines waked by each turbine:
https://github.com/NREL/floris/blob/ef4934ec7feb7afd2615772d364a1eaa28db93e9/floris/simulation/flow_field.py#L724

But this method wouldn't be perfect either, as you pointed out. 

- Evaluating FLORIS would be slower, but I believe it gets evaluated anyway to get the initial wind farm power for normalizing the objective function, so could use those results.
- That's a good point about situations where wake interactions might only appear when there are yaw offsets for some turbines, so this method wouldn't catch them. 
- The method for tracking upstream/downstream turbines only gets executed for two different turbulence models ("crespo_hernandez" and "ishihara_qian"). They would be used the majority of the time, but still the method might not work if another turbulence model is used.

Given these issues, I think it could go either way. Your simple wake model approach might be the better way to go. But in that case, I think it would be good to provide some documentation on what conditions it is valid for and when a user would want to change the default slope. For example, saying that the default slope should handle Gaussian wakes with TI values up to X% using the default Gaussian wake model parameters.

Reply:
Also, should this go in the base optimization folder since it could be used by other non-scipy methods in the future?

Reply:
Yeah good point... It could even be used by other algorithms other than yaw optimization. I am using it in `floris_scada_analysis` to determine which turbines are upstream for a particular wind direction, to which I can then normalize a turbine's power production. Since it is only used in the scipy controls_optimization code for now, I'll leave it here for the time being.